### PR TITLE
storage: shrink direct sequence element ids from `16bytes` to `12bytes`

### PR DESCRIPTION
**Pull request by feds01**
Any thoughts on how to remove a bunch of type casts everywhere?

### CODE REVIEW

**Comment by kontheocharis** [comment]
We should never be able to provide an index explicitly; I believe doing so is unsafe behaviour. We need to find a way to mark it unsafe and change the relevant code that explicitly constructs direct sequence element IDs so that it doesn't handle indices directly.

**Review by feds01 — COMMENTED** [review]
I guess this needs to be decided within the specification, what is the maximum index that can be specified. `usize` seems like the logical choice because it then depends on the platform, however it is always unrealistic (if not impossible to actually create a structure that ever reaches any where near such an index). 

My proposal would be that it is decided by this algorithm:
```rs
min(machine_integer_size(), u32)
```

**Review by feds01 — COMMENTED** [review]
The identifier conversion here is different for `usize` and for `u32`.

This is because internally an `Identifier` is a `u32`, perhaps we should have an explicit function to convert from u32 into a Identifier equivalent?

**Comment by feds01** [comment]
> We should never be able to provide an index explicitly; I believe doing so is unsafe behaviour. We need to find a way to mark it unsafe and change the relevant code that explicitly constructs direct sequence element IDs so that it doesn't handle indices directly.

They have to be constructed somewhere, and sometimes its just mapping things from one type to another. I agree in general that it should be avoided, but without more annoyances. Any suggestions?

File: compiler/hash-lower/src/build/into.rs [hunk 1]
```diff
@@ -477,7 +477,7 @@ impl<'tcx> BodyBuilder<'tcx> {
             .map(|arg| {
                 let name = match arg.target {
                     ParamIndex::Name(name) => name,
-                    ParamIndex::Position(pos) => pos.into(),
+                    ParamIndex::Position(pos) => (pos as usize).into(),
```
Comment (feds01) [inline compiler/hash-lower/src/build/into.rs:480]: same problem here with `Identifier` conversions

File: compiler/hash-exhaustiveness/src/deconstruct.rs [hunk 2]
```diff
@@ -214,7 +214,7 @@ impl<E: ExhaustivenessEnv> fmt::Debug for ExhaustivenessFmtCtx<'_, Deconstructed
                         // currently active.
                         if let DeconstructedCtor::Variant(index) = ctor {
                             let ctors = data_def.borrow().ctors.assert_defined();
-                            let ctor_name = CtorDefId(ctors.elements(), *index).borrow().name;
+                            let ctor_name = CtorDefId::new(ctors.elements(), *index).borrow().name;
```
Comment (kontheocharis) [inline compiler/hash-exhaustiveness/src/deconstruct.rs:217]: Maybe we can provide a method on sequence stores `from_raw_parts` that actually does the assertion that the index is within bounds?

Comment (feds01) [reply compiler/hash-exhaustiveness/src/deconstruct.rs:217]: We could do a `debug_assert!()` in the `new()` constructor?

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**